### PR TITLE
[SHLWAPI][SHLWAPI_APITEST] Expand string in SHLoadIndirectString

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

`shlwapi!SHLoadIndirectString` function seems expanding the environment variable strings.
Split from https://github.com/reactos/reactos/pull/4929.

JIRA issue: [CORE-10667](https://jira.reactos.org/browse/CORE-10667)

## Proposed changes

- `shlwapi!SHLoadIndirectString` expands the environmental strings if the first character was `'@'`.
- Implement `SHLoadRegUIStringA` function.

## TODO

- [x] Do tests.

### CODE REVIEW

**Comment by HBelusca** [comment]
> "TODO: Do tests."

Can this be tested with an apitest?

**Comment by katahiromz** [comment]
WinXP:
WinXP
Successful.

Win2k3:
Win2k3
Successful.

Win10:
Win10
Successful.

**Comment by katahiromz** [comment]
AFTER:
after
Successful.

File: dll/win32/shlwapi/ordinal.c [hunk 1]
```diff
@@ -4121,6 +4121,32 @@ BOOL WINAPI IsOS(DWORD feature)
     return FALSE;
 }
 
+#ifdef __REACTOS__
+/*************************************************************************
+ * @  [SHLWAPI.438]
+ */
+HRESULT WINAPI SHLoadRegUIStringA(HKEY hkey, LPCSTR value, LPSTR buf, DWORD size)
```
Comment (learn-more) [inline dll/win32/shlwapi/ordinal.c:4128]: Usually the A version does do some conversion and calls the W version,
but this works fine for me as well.

File: dll/win32/shlwapi/ordinal.c [hunk 2]
```diff
@@ -4121,6 +4121,32 @@ BOOL WINAPI IsOS(DWORD feature)
...
+
+    if (RegQueryValueExW(hkey, valueW, NULL, NULL, (LPBYTE)bufferW, &sz) != ERROR_SUCCESS)
+        return E_FAIL;
+
+    hr = SHLoadIndirectString(bufferW, bufferW, ARRAY_SIZE(bufferW), NULL);
+    if (FAILED(hr))
+        return hr;
+
+    WideCharToMultiByte(CP_ACP, 0, bufferW, -1, buf, size, NULL, NULL);
+    buf[size - 1] = 0; /* Avoid buffer overrun */
```
Comment (learn-more) [inline dll/win32/shlwapi/ordinal.c:4145]: Does the windows version do this?
I _assume_ that it always ensure the thing is zero terminate, but I doubt it also unconditionally writes at the end of the buffer?

File: dll/win32/shlwapi/string.c [hunk 3]
```diff
@@ -2878,6 +2878,9 @@ HRESULT WINAPI SHLoadIndirectString(LPCWSTR src, LPWSTR dst, UINT dst_len, void
     WCHAR *dllname = NULL;
     HMODULE hmod = NULL;
     HRESULT hr = E_FAIL;
+#ifdef __REACTOS__
+    WCHAR szExpanded[MAX_PATH];
```
Comment (learn-more) [inline dll/win32/shlwapi/string.c:2882]: I am not sure if `MAX_PATH` will be long enough here.

File: modules/rostests/apitests/shlwapi/CMakeLists.txt [hunk 4]
```diff
@@ -23,6 +24,6 @@ add_rc_deps(testdata.rc ${CMAKE_CURRENT_BINARY_DIR}/shlwapi_resource_dll/shlwapi
 add_executable(shlwapi_apitest ${SOURCE})
 set_module_type(shlwapi_apitest win32cui)
 target_link_libraries(shlwapi_apitest ${PSEH_LIB})
-add_importlibs(shlwapi_apitest shlwapi user32 msvcrt kernel32)
+add_importlibs(shlwapi_apitest shlwapi user32 msvcrt advapi32 kernel32)
```
Comment (learn-more) [inline modules/rostests/apitests/shlwapi/CMakeLists.txt:27]: Please move `advapi32` left of `msvcrt`

File: modules/rostests/apitests/shlwapi/SHLoadRegUIString.c [hunk 5]
```diff
@@ -0,0 +1,96 @@
+/*
+ * PROJECT:     ReactOS api tests
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Tests for SHLoadRegUIStringA/W
+ * COPYRIGHT:   Copyright 2023 Katayama Hirofumi MZ (katayama.hirofumi.mz@gmail.com)
```
Comment (learn-more) [inline modules/rostests/apitests/shlwapi/SHLoadRegUIString.c:5]:
```suggestion
 * COPYRIGHT:   Copyright 2023 Katayama Hirofumi MZ <katayama.hirofumi.mz@gmail.com>
```

**Review by learn-more — APPROVED** [review]
Some minor comments.

File: dll/win32/shlwapi/ordinal.c (hunk 2 above)
Comment (katahiromz) [reply dll/win32/shlwapi/ordinal.c:4145]: > Does the windows version do this?

Surely.

> I doubt it also unconditionally writes at the end of the buffer?
`buffer` has its length of `size`. if `size` is non-zero, `&buf[size - 1]` is in `buf`.

File: modules/rostests/apitests/shlwapi/CMakeLists.txt (hunk 4 above)
Comment (katahiromz) [reply modules/rostests/apitests/shlwapi/CMakeLists.txt]: OK.

File: modules/rostests/apitests/shlwapi/SHLoadRegUIString.c (hunk 5 above)
Comment (katahiromz) [reply modules/rostests/apitests/shlwapi/SHLoadRegUIString.c]: OK.

File: dll/win32/shlwapi/string.c (hunk 3 above)
Comment (katahiromz) [reply dll/win32/shlwapi/string.c]: `512` is enough.

File: dll/win32/shlwapi/ordinal.c (hunk 2 above)
Comment (katahiromz) [reply dll/win32/shlwapi/ordinal.c]: Zero check added.

File: dll/win32/shlwapi/ordinal.c [hunk 6]
```diff
@@ -4121,6 +4121,33 @@ BOOL WINAPI IsOS(DWORD feature)
...
 }
 
+#ifdef __REACTOS__
+/*************************************************************************
+ * @  [SHLWAPI.438]
+ */
+HRESULT WINAPI SHLoadRegUIStringA(HKEY hkey, LPCSTR value, LPSTR buf, DWORD size)
+{
+    WCHAR valueW[MAX_PATH], bufferW[MAX_PATH];
+    DWORD sz = ARRAY_SIZE(bufferW) * sizeof(CHAR);
```
Comment (HBelusca) [inline dll/win32/shlwapi/ordinal.c:4131]: Use "dwSize", or just "size" instead. (sz means "string zero terminated", usually)

Comment (katahiromz) [reply dll/win32/shlwapi/ordinal.c]: `dwSize` is used.

File: dll/win32/shlwapi/ordinal.c [hunk 7]
```diff
@@ -4121,6 +4121,33 @@ BOOL WINAPI IsOS(DWORD feature)
...
+    if (FAILED(hr))
+        return hr;
+
+    WideCharToMultiByte(CP_ACP, 0, bufferW, -1, buf, size, NULL, NULL);
+    if (size > 0)
+        buf[size - 1] = 0; /* Avoid buffer overrun */
+    return S_OK;
+}
+
+#endif
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shlwapi/ordinal.c]: Moved.

File: modules/rostests/apitests/shlwapi/SHLoadRegUIString.c [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    hSHLWAPI = GetModuleHandleW(L"shlwapi");
+    pSHLoadRegUIStringA = (FN_SHLoadRegUIStringA)GetProcAddress(hSHLWAPI, (LPCSTR)438);
+    pSHLoadRegUIStringW = (FN_SHLoadRegUIStringW)GetProcAddress(hSHLWAPI, (LPCSTR)439);
+    if (!pSHLoadRegUIStringA || !pSHLoadRegUIStringW)
+    {
+        skip("No procedure found\n");
+        return;
+    }
+
+    if (!extract_resource(L"SHLoadRegUIString.dll", MAKEINTRESOURCEW(101)))
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply modules/rostests/apitests/shlwapi/SHLoadRegUIString.c:68]: Resource.

[the remark this answers is not in the document]
Comment (HBelusca) [reply modules/rostests/apitests/shlwapi/SHLoadRegUIString.c:68]: Ah ok, it creates the file :+1:

**Review by HBelusca — APPROVED** [review]

**Pull request merged**